File: lib/cyanea/align.ex
defmodule Cyanea.Align do
  @moduledoc "Pairwise and multiple sequence alignment."

  import Cyanea.NifHelper
  alias Cyanea.Native

  # ===========================================================================
  # Pairwise DNA
  # ===========================================================================

  @doc """
  Align two DNA sequences.

  ## Options

    * `:mode` - `:local` (default), `:global`, or `:semiglobal`
    * `:match` - match score (triggers custom scoring when any scoring opt is present)
    * `:mismatch` - mismatch penalty
    * `:gap_open` - gap open penalty
    * `:gap_extend` - gap extend penalty

  When no scoring options are given, uses defaults (+2/-1/-5/-2).
  """
  @spec dna(binary(), binary(), keyword()) :: {:ok, struct()} | {:error, term()}
  def dna(query, target, opts \\ []) when is_binary(query) and is_binary(target) do
    mode = mode_string(Keyword.get(opts, :mode, :local))

    if has_scoring_opts?(opts) do
      match = Keyword.get(opts, :match, 2)
      mismatch = Keyword.get(opts, :mismatch, -1)
      gap_open = Keyword.get(opts, :gap_open, -5)
      gap_extend = Keyword.get(opts, :gap_extend, -2)
      nif_call(fn -> Native.align_dna_custom(query, target, mode, match, mismatch, gap_open, gap_extend) end)
    else
      nif_call(fn -> Native.align_dna(query, target, mode) end)
    end
  end

  defp has_scoring_opts?(opts) do
    Keyword.has_key?(opts, :match) or Keyword.has_key?(opts, :mismatch) or
      Keyword.has_key?(opts, :gap_open) or Keyword.has_key?(opts, :gap_extend)
  end

  # ===========================================================================
  # Pairwise protein
  # ===========================================================================

  @doc """
  Align two protein sequences.

  ## Options

    * `:mode` - `:global` (default), `:local`, or `:semiglobal`
    * `:matrix` - `:blosum62` (default), `:blosum45`, `:blosum80`, or `:pam250`

  """
  @spec protein(binary(), binary(), keyword()) :: {:ok, struct()} | {:error, term()}
  def protein(query, target, opts \\ []) when is_binary(query) and is_binary(target) do
    mode = mode_string(Keyword.get(opts, :mode, :global))
    matrix = matrix_string(Keyword.get(opts, :matrix, :blosum62))
    nif_call(fn -> Native.align_protein(query, target, mode, matrix) end)
  end

  # ===========================================================================
  # Batch
  # ===========================================================================

  @doc """
  Batch-align a list of `{query, target}` DNA pairs (DirtyCpu scheduler).

  ## Options

    * `:mode` - `:local` (default), `:global`, or `:semiglobal`

  """
  @spec batch(list(), keyword()) :: {:ok, list()} | {:error, term()}
  def batch(pairs, opts \\ []) when is_list(pairs) do
    mode = mode_string(Keyword.get(opts, :mode, :local))
    nif_call(fn -> Native.align_batch_dna(pairs, mode) end)
  end

  # ===========================================================================
  # MSA
  # ===========================================================================

  @doc """
  Progressive multiple sequence alignment.

  ## Options

    * `:mode` - `:dna` (default) or `:protein`

  """
  @spec msa(list(), keyword()) :: {:ok, struct()} | {:error, term()}
  def msa(sequences, opts \\ []) when is_list(sequences) do
    mode = msa_mode_string(Keyword.get(opts, :mode, :dna))
    nif_call(fn -> Native.progressive_msa(sequences, mode) end)
  end

  # ===========================================================================
  # Banded alignment
  # ===========================================================================

  @doc """
  Banded DNA alignment. Restricts DP to diagonal band of `2*bandwidth+1`.

  ## Options

    * `:mode` - `:global` (default), `:local`, or `:semiglobal`
    * `:bandwidth` - band half-width (default: 50)

  """
  @spec banded(binary(), binary(), keyword()) :: {:ok, struct()} | {:error, term()}
  def banded(query, target, opts \\ []) when is_binary(query) and is_binary(target) do
    mode = mode_string(Keyword.get(opts, :mode, :global))
    bandwidth = Keyword.get(opts, :bandwidth, 50)
    nif_call(fn -> Native.banded_align_dna(query, target, mode, bandwidth) end)
  end

  @doc """
  Banded alignment score only (no traceback, less memory).

  ## Options

    * `:mode` - `:global` (default), `:local`, or `:semiglobal`
    * `:bandwidth` - band half-width (default: 50)

  """
  @spec banded_score(binary(), binary(), keyword()) :: {:ok, integer()} | {:error, term()}
  def banded_score(query, target, opts \\ []) when is_binary(query) and is_binary(target) do
    mode = mode_string(Keyword.get(opts, :mode, :global))
    bandwidth = Keyword.get(opts, :bandwidth, 50)
    nif_call(fn -> Native.banded_score_only(query, target, mode, bandwidth) end)
  end

  # ===========================================================================
  # POA consensus
  # ===========================================================================

  @doc "Compute consensus from multiple sequences using Partial Order Alignment."
  @spec consensus(list()) :: {:ok, binary()} | {:error, term()}
  def consensus(sequences) when is_list(sequences),
    do: nif_call(fn -> Native.poa_consensus(sequences) end)
end

File: native/cyanea_native/src/align.rs
//! cyanea-align NIFs — Pairwise alignment, batch, MSA, banded, POA.

use crate::bridge::*;
use crate::to_nif_error;

// ===========================================================================
// Helpers
// ===========================================================================

pub(crate) fn parse_alignment_mode(mode: &str) -> Result<cyanea_align::AlignmentMode, String> {
    match mode {
        "local" => Ok(cyanea_align::AlignmentMode::Local),
        "global" => Ok(cyanea_align::AlignmentMode::Global),
        "semiglobal" => Ok(cyanea_align::AlignmentMode::SemiGlobal),
        _ => Err(format!("unknown alignment mode: {mode} (expected local, global, or semiglobal)")),
    }
}

fn parse_substitution_matrix(name: &str) -> Result<cyanea_align::SubstitutionMatrix, String> {
    match name {
        "blosum62" => Ok(cyanea_align::SubstitutionMatrix::blosum62()),
        "blosum45" => Ok(cyanea_align::SubstitutionMatrix::blosum45()),
        "blosum80" => Ok(cyanea_align::SubstitutionMatrix::blosum80()),
        "pam250" => Ok(cyanea_align::SubstitutionMatrix::pam250()),
        _ => Err(format!("unknown substitution matrix: {name} (expected blosum62, blosum45, blosum80, or pam250)")),
    }
}

// ===========================================================================
// Existing NIFs
// ===========================================================================

#[rustler::nif]
pub fn align_dna(query: Vec<u8>, target: Vec<u8>, mode: String) -> Result<AlignmentResultNif, String> {
    let mode = parse_alignment_mode(&mode)?;
    let scoring = cyanea_align::ScoringScheme::Simple(cyanea_align::ScoringMatrix::dna_default());
    cyanea_align::align(&query, &target, mode, &scoring)
        .map(AlignmentResultNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif]
pub fn align_dna_custom(
    query: Vec<u8>,
    target: Vec<u8>,
    mode: String,
    match_score: i32,
    mismatch_score: i32,
    gap_open: i32,
    gap_extend: i32,
) -> Result<AlignmentResultNif, String> {
    let mode = parse_alignment_mode(&mode)?;
    let matrix = cyanea_align::ScoringMatrix::new(match_score, mismatch_score, gap_open, gap_extend)
        .map_err(to_nif_error)?;
    let scoring = cyanea_align::ScoringScheme::Simple(matrix);
    cyanea_align::align(&query, &target, mode, &scoring)
        .map(AlignmentResultNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif]
pub fn align_protein(
    query: Vec<u8>,
    target: Vec<u8>,
    mode: String,
    matrix: String,
) -> Result<AlignmentResultNif, String> {
    let mode = parse_alignment_mode(&mode)?;
    let sub_matrix = parse_substitution_matrix(&matrix)?;
    let scoring = cyanea_align::ScoringScheme::Substitution(sub_matrix);
    cyanea_align::align(&query, &target, mode, &scoring)
        .map(AlignmentResultNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn align_batch_dna(
    pairs: Vec<(Vec<u8>, Vec<u8>)>,
    mode: String,
) -> Result<Vec<AlignmentResultNif>, String> {
    let mode = parse_alignment_mode(&mode)?;
    let scoring = cyanea_align::ScoringScheme::Simple(cyanea_align::ScoringMatrix::dna_default());
    let refs: Vec<(&[u8], &[u8])> = pairs.iter().map(|(q, t)| (q.as_slice(), t.as_slice())).collect();
    cyanea_align::align_batch(&refs, mode, &scoring)
        .map(|results| results.into_iter().map(AlignmentResultNif::from).collect())
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn progressive_msa(sequences: Vec<Vec<u8>>, mode: String) -> Result<MsaResultNif, String> {
    let refs: Vec<&[u8]> = sequences.iter().map(|s| s.as_slice()).collect();
    let scoring = match mode.as_str() {
        "dna" => {
            cyanea_align::ScoringScheme::Simple(cyanea_align::ScoringMatrix::dna_default())
        }
        "protein" => cyanea_align::ScoringScheme::Substitution(
            cyanea_align::SubstitutionMatrix::blosum62(),
        ),
        _ => return Err(format!("unknown MSA mode: {mode} (expected dna or protein)")),
    };
    let result = cyanea_align::msa::progressive_msa(&refs, &scoring).map_err(to_nif_error)?;
    let n_sequences = result.n_sequences();
    let n_columns = result.n_columns;
    let conservation = result.conservation();
    Ok(MsaResultNif {
        aligned: result.aligned,
        n_sequences,
        n_columns,
        conservation,
    })
}

// ===========================================================================
// New NIFs
// ===========================================================================

#[rustler::nif]
pub fn banded_align_dna(
    query: Vec<u8>,
    target: Vec<u8>,
    mode: String,
    bandwidth: usize,
) -> Result<AlignmentResultNif, String> {
    let mode = parse_alignment_mode(&mode)?;
    let scoring = cyanea_align::ScoringScheme::Simple(cyanea_align::ScoringMatrix::dna_default());
    let result = match mode {
        cyanea_align::AlignmentMode::Global => {
            cyanea_align::simd::banded_nw(&query, &target, &scoring, bandwidth)
        }
        cyanea_align::AlignmentMode::Local => {
            cyanea_align::simd::banded_sw(&query, &target, &scoring, bandwidth)
        }
        cyanea_align::AlignmentMode::SemiGlobal => {
            cyanea_align::simd::banded_semi_global(&query, &target, &scoring, bandwidth)
        }
    };
    result.map(AlignmentResultNif::from).map_err(to_nif_error)
}

#[rustler::nif]
pub fn banded_score_only(
    query: Vec<u8>,
    target: Vec<u8>,
    mode: String,
    bandwidth: usize,
) -> Result<i32, String> {
    let mode = parse_alignment_mode(&mode)?;
    let scoring = cyanea_align::ScoringScheme::Simple(cyanea_align::ScoringMatrix::dna_default());
    cyanea_align::simd::banded_score_only(&query, &target, &scoring, bandwidth, mode)
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn poa_consensus(sequences: Vec<Vec<u8>>) -> Result<Vec<u8>, String> {
    if sequences.is_empty() {
        return Err("at least one sequence required".into());
    }
    let scoring = cyanea_align::poa::PoaScoring {
        match_score: 2,
        mismatch_score: -1,
        gap_score: -2,
    };
    let mut graph = cyanea_align::poa::PoaGraph::from_sequence(&sequences[0]);
    for seq in &sequences[1..] {
        graph.add_sequence(seq, &scoring).map_err(to_nif_error)?;
    }
    Ok(graph.consensus())
}
